### DISCUSSION TITLE
Inject settings in to HttpClientFactory extension

### DISCUSSION THREAD

**Discussion by VissersR** [canton7/RestEase#247] [General]
The HTTPClientFactory extension is a nice way to add restease clients to the DI process, but at the moment it is not possible to inject the settings or alternative serializers. I've been playing around with the source and it is not that difficult to add an overload that takes a delegate with a service provider without changing anything substantial. 
That is besides one thing, the requestModifier. I haven't found a clean way yet to deal with nulls for that, as the AddHttpMessageHandler call on the HTTPClientBuilder does not check for nulls when taking a delegate and in the current extension structure that would be the only place to resolve wat is passed in.

I have managed to work around almost all the limitations be using ConfigureHttpClient and ConfigurePrimaryHttpMessageHandler, But I still need the RestClientConfigurer method to set the serializer/deserializer and QueryStringBuilder, and it would be nicer to just have one nice configuration section.

Is this something that is planned? Or something you would accept a pullrequest/proposal for?

**Comment by canton7** [comment]
Can't you use the method described in the [second snippet in the docs](https://github.com/canton7/RestEase#using-httpclientfactory)?

```cs
services.AddRestEaseClient<ISomeApi>("https://api.example.com", new()
{
    RestClientConfigurer = client =>
    {
        client.RequestPathParamSerializer = new StringEnumRequestPathParamSerializer(),
        // Other serializers, settings, etc?
    },
});
```

**Reply by VissersR** [reply]
No because you don't have a serviceprovider there. 

```cs
    services.AddRestEaseClient<T>(new AddRestEaseClientOptions<T>()
    {
        RestClientConfigurer = client =>
        {
            client.ResponseDeserializer = // Get from serviceprovider
            client.RequestBodySerializer = // Get from serviceprovider
            client.QueryStringBuilder = // Get from serviceprovider
        },
    })
    .ConfigureHttpClient((sp, client) =>
    {
        var endpoint = sp.GetRequiredService<TSettings>().Endpoint;
        client.BaseAddress = new Uri(endpoint);
    })
    .ConfigurePrimaryHttpMessageHandler(sp => 
    new CustomHandler<TSettings>(
        sp.GetRequiredService<ILogger>(), 
        sp.GetRequiredService<IOauthService>(), 
        sp.GetRequiredService<TSettings>()));
    
    return services;
````
I roughly do this at the moment.

As you can see, the ConfigureHttpClient and ConfigurePrimaryHttpMessageHandler have a delegate that provide a ServiceProvider, so what I would like is a Func<IServiceProvider, AddRestEaseClientOptions>> settingsAction overload.

Currently RestEase resolves all of the options in a way that a service provider is available.

```cs
        private static IHttpClientBuilder AddRestEaseClientCore(
            this IHttpClientBuilder httpClientBuilder,
            Type restEaseType,
            Func<HttpClient, object> factory,
            RequestModifier? requestModifier)
        {
            if (httpClientBuilder is null)
                throw new ArgumentNullException(nameof(httpClientBuilder));

            // See https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs
            httpClientBuilder.Services.AddTransient(restEaseType, **serviceProvider** =>
            {
                var httpClientFactory = serviceProvider.GetRequiredService<IHttpClientFactory>();
                var httpClient = httpClientFactory.CreateClient(httpClientBuilder.Name);
                return factory(httpClient);
            });

            if (requestModifier != null)
            {
                // Also has a serviceprovider delegate overload
                httpClientBuilder.AddHttpMessageHandler(() => new ModifyingClientHttpHandler(requestModifier));
            }

            return httpClientBuilder;
        }
```
But AddHttpMessageHandler does not do null checks, and because it would then be a delegate you don't know if it is null until it eventually is resolved out the scope that would prevent passing null.

**Reply by canton7** [reply]
Aha, I follow what you're after. I'll try and put something in place this weekend

**Comment by canton7** [comment]
Turns out, coming up with a nice API for this, which doesn't break backwards compat, is irritating. I'll have to have a think about how best to do this...

**Reply by VissersR** [reply]
Yeah, thats what I found out as well otherwise I would have suggested a nice solution that fits the current design. 
If I have epiphany I'll be sure to let you know.

**Reply by canton7** [reply]
Yeah, I've tried:

 * `AddRestEaseClient<T>("url", services => new() ...)`
 * `AddRestEaseClient<T>("url", new().WithResponseDeserializer(services => ...))`
 * `AddRestEaseClient<T>("url", builder => builder.WithResponseDeserializer(services => ...))`

The main problem is that I've advocated using `new()` to instantiate the `AddRestEaseClientOptions`, and adding any overload which takes any other object types causes an ambiguity.

The only remaining option I can see is to add additional properties to `AddRestEaseClientOptions`, e.g. `ResponseDeserializerWithServices`, which is a bit ugly...
